spec/01-unit/26-tracing_spec.lua

```diff
@@ -0,0 +1,238 @@
+      local span = c_tracer.start_span("meow")
+      assert.error(function() span:set_attribute("key1") end)
+      assert.error(function() span:set_attribute("key1", function() end) end)
+      assert.error(function() span:set_attribute(123, 123) end)
+    end)
+
+    it("fails add_event", function ()
+      local span = c_tracer.start_span("meow")
+      assert.error(function() span:set_attribute("key1") end)
+      assert.error(function() span:set_attribute("key1", function() end) end)
+      assert.error(function() span:set_attribute(123, 123) end)
+    end)
+
+    it("child spans", function ()
+      local root_span = c_tracer.start_span("parent")
+      c_tracer.set_active_span(root_span)
+      local child_span = c_tracer.start_span("child")
+
+      assert.same(root_span.span_id, child_span.parent_id)
+
+      local second_child_span = c_tracer.start_span("child2")
+      assert.same(root_span.span_id, second_child_span.parent_id)
+      assert.are_not.same(child_span.span_id, second_child_span.parent_id)
+
+      c_tracer.set_active_span(child_span)
+      local third_child_span = c_tracer.start_span("child2")
+      assert.same(child_span.span_id, third_child_span.parent_id)
+    end)
+
+    it("cascade spans", function ()
+      local root_span = c_tracer.start_span("root")
+      c_tracer.set_active_span(root_span)
+
+      assert.same(root_span, c_tracer.active_span())
+
+      local level1_span = c_tracer.start_span("level1")
+      assert.same(root_span, level1_span.parent)
+
+      c_tracer.set_active_span(level1_span)
+      assert.same(level1_span, c_tracer.active_span())
+
+      local level2_span = c_tracer.start_span("level2")
+      assert.same(level1_span, level2_span.parent)
+
+      local level21_span = c_tracer.start_span("level2.1")
+      assert.same(level1_span, level21_span.parent)
+      level21_span:finish()
+
+      c_tracer.set_active_span(level2_span)
+      assert.same(level2_span, c_tracer.active_span())
+
+      local level3_span = c_tracer.start_span("level3")
+      assert.same(level2_span, level3_span.parent)
+      level3_span:finish()
+
+      level2_span:finish()
+      assert.same(level1_span, c_tracer.active_span())
+
+      level1_span:finish()
+      assert.same(root_span, c_tracer.active_span())
+    end)
+
+    it("access span table after finished", function ()
+      local span = c_tracer.start_span("meow")
+      span:finish()
+      assert.has_no.error(function () span:finish() end)
+    end)
+
+    it("ends active span", function ()
+      local span = c_tracer.start_span("meow")
+      c_tracer.set_active_span(span)
+
+      -- create sub spans
+      local sub = c_tracer.start_span("sub")
+      sub:finish()
+      sub = c_tracer.start_span("sub")
+      sub:finish()
+
+      local active_span = c_tracer.active_span()
+      assert.same(span, active_span)
+      assert.has_no.error(function () active_span:finish() end)
+
+      assert.same(span, active_span)
```

Comment:
I'm not sure what this second `assert.same(span, active_span)` is testing. Can you clarify?

Reply:
to make sure the span is still same after `finish`

Reply:
In this case, `span` and `active_span` are two references to the same object, since `tracer.active_span()` and `tracer.set_active_span(span)` don't do any copying. Inserting `assert.equal(span, active_span)` at line `231` confirms this):

```diff
@@ -229,6 +229,7 @@ describe("Tracer PDK", function()

       local active_span = c_tracer.active_span()
       assert.same(span, active_span)
+      assert.equal(span, active_span)
       assert.has_no.error(function () active_span:finish() end)

       -- make sure the span is still same after finish
```

This means that the final `assert.same(span, active_span)` does not test anything; it will always succeed regardless of any unintended side effects of `active_span:finish()`:

```lua
local span = { a = 1 }
local active_span = span
assert.same(span, active_span)

-- do something that has a side effect on span/active_span
span.a = 5 
active_span.b = 2

-- this test will always succeed; it is equivalent to:
-- `assert.same({ a = 5, b = 2 }, { a = 5, b = 2 })`
assert.same(span, active_span)
```

Reply:
Oh, my fault, thanks for pointing it out, this won't be testing the functionality, my purpose was to test if the original span is not cleared (e.g. `table.clear`) and still be accessible for callers.

I've changed the test to 
```lua
      assert.has_no.error(function () active_span:finish() end)

      -- span's property is still accessible
      assert.same("meow", active_span.name)
```